src/libsyntax/feature_gate.rs

```diff
@@ -1295,6 +1298,16 @@ fn contains_novel_literal(item: &ast::MetaItem) -> bool {
     }
 }
 
+impl<'a> PostExpansionVisitor<'a> {
+    fn whole_crate_feature_gates(&mut self) {
+        for &span in &*self.context.parse_sess.non_modrs_mods.borrow() {
+            gate_feature_post!(
+                &self, non_modrs_mods, span,
+                "mod declarations in non-mod.rs files are experimental");
```

Comment:
maybe we want to add some extra notes here, given that this is a common source of confusion? Something like:

> on stable builds, rename `foo.rs` to `foo/mod.rs` to enable submodules

not sure how hard that would be to do